Introduce message queue to better correlate requests and responses

fbiville opened the pull request:
This is currently only applied to Bolt 5.x, but can be ported to
older protocol connections as well.

This removes some fragile heuristics on response correlation
(see e.g. success#isResetResponse).

This also paves the way for a cleaner implementation of re-auth.

Big kudos to @robsdedude for guiding through this, as he already
implemented a similar approach in the Python driver.

fbiville committed:
Rephrase internal stream handling documentation

fbiville committed:
Introduce message queue to better correlate requests and responses

This is currently only applied to Bolt 5.x, but can be ported to
older protocol connections as well.

This removes some fragile heuristics on response correlation
(see e.g. success#isResetResponse).

This also paves the way for a cleaner implementation of re-auth.

Big kudos to @robsdedude for guiding through the design principles
of this solution.

fbiville committed:
Make Reset exhaust callback queue before sending RESET

Couple more fixes:

 - make PULL response handler enqueue itself first (not last)
 - fix infinite loop in stream pause

fbiville committed:
Remove nonsensical response's handlers for RESET

fbiville committed:
Polish

fbiville committed:
Do not assume current stream's qid is the last one seen

fbiville committed:
Remove buggy uses of detach

Stream may be pulling results, but it's not necessarily the current
one

fbiville committed:
Fix TestKit tests

This fixes the behavior when more than one result stream is being
pulled.

fbiville committed:
Polish

robsdedude commented on neo4j/internal/bolt/message_queue.go:
```diff
@@ -0,0 +1,206 @@
+	q.protocolVersion = protocolVersion
+	q.out.appendHello(hello)
+	q.enqueueCallback(helloHandler)
+	if q.protocolVersion.greaterThanOrEqual(5, 1) {
+		q.out.appendLogon(token)
+		q.enqueueCallback(logonHandler)
+	}
```
I think it would make for a cleaner separation of concerns if the queue was more (ideally fully) protocol version agnostic.

In this particular case there could be an `appendHello` method on the queue and `bolt.Connect` is responsible for deciding whether logon needs to be enqueued or not.

fbiville replied:
I agree and had the same thought yesterday evening!
Keeping protocol version-based logic here would make things too complicated (esp. for bolt4 and ROUTE)

fbiville replied:
Fixed in 294952d4d188f3e2c4b90be8158ea02414e676b4

robsdedude commented on neo4j/internal/bolt/bolt5.go:
```diff
@@ -427,29 +386,31 @@ func (b *bolt5) discardStream(ctx context.Context) {
+		if stream.sum != nil || stream.err != nil {
+			return
+		}
+		if stream.endOfBatch && discarded {
```
I think this check is not sufficient. Try writing a test to verify. This is the scenario I'm thinking of:
 1. Make a record stream with 4 records
 2. set fetch size to 2
 3. call `bolt.Next` 3 times (add another test with calling it 4 times)
 4. call `bolt.Discard`.
 
 The problem I suspect: there is nothing left to be discarded; everything has been `PULL`ed already, but the driver will enqueue a `DISCARD` anyway.

fbiville replied:
if everything is pulled, then `stream.sum` should not be nil and will return earlier (see condition just above)?

fbiville replied:
proven in 6ad32ecbd7d9d75434a1fc70e2cef1efbdff5cf6

robsdedude replied:
Ah yes! I failed to realize the interplay with the `receiveAll` and `if stream.sum != nil || stream.err != nil` above. :ok_hand:

robsdedude commented on neo4j/internal/bolt/message_queue.go:
```diff
@@ -0,0 +1,206 @@
+	q.enqueueCallback(handler)
+}
+
+func (q *messageQueue) appendRun(cypher string, params, meta map[string]any, fetchSize int,
```
Similar to my comment on `appendHello`, I think `appendRun` should not be responsible for pipelining `PULL`, but bolt should probably do this explicitly.

fbiville replied:
Fixed in 50fddd2123cf63c1a23fe58bec26223b1ae84f68

robsdedude commented on neo4j/test-integration/dbconn_test.go:
```diff
@@ -22,6 +22,8 @@ package test_integration
 	"context"
 	"crypto/rand"
+	"fmt"
+	"github.com/neo4j/neo4j-go-driver/v5/neo4j/db"
```
Why did the driver dependency move up here? :detective:

fbiville replied:
something something leftover something

fbiville replied:
I don't think this file changes very often, so maybe `go fmt` behavior changed a bit since last time?

robsdedude commented on neo4j/test-integration/dbconn_test.go:
```diff
@@ -101,382 +102,390 @@ func TestConnectionConformance(outer *testing.T) {
+			//			idb.TxConfig{Mode: idb.WriteMode})
+			//		AssertNoError(t, err)
+			//	},
+			//},
+			//// Connection is in transaction
+			//{
+			//	name: "Set connection in transaction mode",
+			//	fun: func(t *testing.T, c idb.Connection) {
+			//		tx, _ := c.TxBegin(context.Background(),
+			//			idb.TxConfig{Mode: idb.WriteMode})
+			//		_, err := c.RunTx(context.Background(), tx,
+			//			idb.Command{Cypher: "UNWIND [1] AS n RETURN n"})
+			//		AssertNoError(t, err)
+			//	},
+			//},
+			//// Really big stream streaming
+			//{
+			//	name: "Streaming big stream (autocommit)",
+			//	fun: func(t *testing.T, c idb.Connection) {
+			//		_, err := c.Run(context.Background(),
+			//			idb.Command{Cypher: "UNWIND RANGE (0, " +
+			//				"1000000) AS x RETURN x"}, idb.TxConfig{Mode: idb.
+			//				ReadMode})
+			//		AssertNoError(t, err)
+			//	},
+			//},
+			//// Big nested streams in an uncommitted tx
+			//{
+			//	name: "Streaming big stream (explicit tx)",
+			//	fun: func(t *testing.T, c idb.Connection) {
+			//		tx, _ := c.TxBegin(context.Background(),
+			//			idb.TxConfig{Mode: idb.WriteMode})
+			//		_, err := c.RunTx(context.Background(), tx,
+			//			idb.Command{Cypher: "UNWIND RANGE (0, 1000000) AS n RETURN n"})
+			//		AssertNoError(t, err)
+			//		_, err = c.RunTx(context.Background(), tx,
+			//			idb.Command{Cypher: "UNWIND RANGE (0, 1000000) AS n RETURN n"})
+			//		AssertNoError(t, err)
+			//	},
+			//},
```
:eyes:

fbiville committed:
Reinstate tests 🙈

fbiville committed:
Fix tfirst and tlast logging of success message

fbiville committed:
Remove protocol logic from message queue

fbiville committed:
Remove PULL pipelining from message queue

fbiville committed:
Make integration test re-runnable

Rewrite assertions so that they do not assume an empty state at
the beginning of the test execution

fbiville committed:
Make sure discarding works

robsdedude reviewed (approved):
:up:

fbiville committed:
Add unit tests+polish
